src/main/java/org/xbill/DNS/DNSSession.java

```diff
@@ -0,0 +1,129 @@
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import org.xbill.DNS.exceptions.AdditionalDetail;
+import org.xbill.DNS.exceptions.InvalidZoneDataException;
+import org.xbill.DNS.exceptions.LookupFailedException;
+import org.xbill.DNS.exceptions.RedirectOverflowException;
+import org.xbill.DNS.exceptions.ServerFailedException;
+
+/**
+ * DNSSession provides facilities to make DNS Queries. A DNSSession is intended to be long lived,
+ * and it's behaviour can be modified by calling the setter methods.
+ */
+public class DNSSession {
+  private static final int DEFAULT_MAX_ITERATIONS = 16;
+
+  private final Resolver resolver;
+  private volatile int maxRedirects = DEFAULT_MAX_ITERATIONS;
+
+  /**
+   * Construct a DNSSession using the provided Resolver to lookup records.
+   *
+   * @param resolver the Resolver instance to backing this DNSSession.
+   */
+  DNSSession(Resolver resolver) {
+    this.resolver = resolver;
+  }
+
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * RedirectOverflowException.
+   *
+   * @param maxRedirects the maximum number of allowed redirections.
+   */
+  public void setMaxRedirects(int maxRedirects) {
+    this.maxRedirects = maxRedirects;
+  }
+
+  /**
+   * Make an asynchronous lookup of the provided name.
+   *
+   * @param name the name to look up.
+   * @param type the type to look up, values should correspond to constants in Type.
+   * @param dclass the class to look up, values should correspond to constants in DClass.
+   * @return A completion stage what will yield the eventual lookup result.
+   */
+  public CompletionStage<LookupResult> lookupAsync(Name name, int type, int dclass) {
+    Record question = Record.newRecord(name, type, dclass);
+    Message query = Message.newQuery(question);
+
+    return resolver.sendAsync(query).thenCompose(this::resolveRedirects);
+  }
+
+  private CompletionStage<LookupResult> resolveRedirects(Message response) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    maybeFollowRedirect(response, 1, future);
+    return future;
+  }
+
+  private void maybeFollowRedirect(
+      Message response, int redirectCount, CompletableFuture<LookupResult> future) {
+    try {
+      if (redirectCount > maxRedirects) {
+        throw new RedirectOverflowException(
+            format("Refusing to follow more than %s redirects", maxRedirects));
+      }
+
+      List<Record> allAnswers = response.getSection(Section.ANSWER);
+      LookupResult result = dealWithEmptyAnswer(allAnswers, response.getHeader().getRcode());
+      if (result != null) {
+        future.complete(result);
+        return;
+      }
+      Record firstAnswer = allAnswers.get(0);
+
+      if (firstAnswer.getType() == Type.DNAME || firstAnswer.getType() == Type.CNAME) {
+        resolver
+            .sendAsync(Message.newQuery(buildRedirectQuery(response)))
```

Comment:
I know this is like the original Lookup class. But is it even legal for a non-recursive resolver to follow CNAMEs if the answer from a recursing resolver is incomplete?

Reply:
I agree. It is strange. I just wanted to make it a drop in replacement. Come to think of it I think that it makes sense to have maxRedirects default to 0 and have that mean to not follow redirects and have Lookup use the old behaviour at least until the next major release.

Reply:
maxRedirects in the existing Lookup class uses that parameter also to follow the chain within a Message, which is fine. Not following CNAMEs by sending more queries to a (recursive) resolvers is something that can be done in a minor version, or changed via an option.

Reply:
I'll look into changing that as a followup to this, but I feel I would like to land this first and implementing the old behaviour first seems like a good idea.